core/commons/che-core-commons-test/src/main/java/org/eclipse/che/commons/test/mockito/answer/WaitingAnswer.java

```diff
@@ -0,0 +1,160 @@
+ *     doAnswer(waitingAnswer).when(someClassUsedInTestedClass).someMethod(eq(param1), eq(param2));
+ *
+ *     // start doing something in a separate thread
+ *     executor.execute(() -> testedClass.doSomething());
+ *     // wait until separate thread call answer to find the moment when critical area is occupied
+ *     // to make test fast wait not more that provided timeout
+ *     waitingAnswer.waitAnswerCall(1, TimeUnit.SECONDS);
+ *
+ *     // when
+ *     try {
+ *         // start doing something in current thread
+ *         testedClass.doSomething()
+ *         // this area should not be reachable until answer is completed!
+ *         fail("Error message");
+ *     } finally {
+ *         // In this case exception can be suppressed
+ *         // Test is simplified to provide clean example
+ *
+ *         // then
+ *         // complete waiting answer
+ *         waitingAnswer.completeAnswer();
+ *         // ensure that someMethod is called only once to confirm that testedClass.doSomething()
+ *         // doesn't call it this particular test
+ *         verify(someClassUsedInTestedClass, timeout(100).times(1)).someMethod(any(), any());
+ *     }
+ * </code></pre>
+ *
+ * @author Alexander Garagatyi
+ */
+public class WaitingAnswer<T> implements Answer<T> {
+
+    private final CountDownLatch answerIsCalledLatch;
+    private final CountDownLatch answerResultIsUnlockedLatch;
+
+    private long     maxWaitingTime;
+    private TimeUnit maxWaitingUnit;
+    private T        result;
+
+    private volatile String error;
+
+    public WaitingAnswer() {
+        this.maxWaitingTime = 1;
+        this.maxWaitingUnit = TimeUnit.SECONDS;
+        this.result = null;
+        this.answerIsCalledLatch = new CountDownLatch(1);
+        this.answerResultIsUnlockedLatch = new CountDownLatch(1);
+        this.error = null;
+    }
+
+    public WaitingAnswer(long maxWaitingTime, TimeUnit maxWaitingUnit) {
+        this();
+        this.maxWaitingTime = maxWaitingTime;
+        this.maxWaitingUnit = maxWaitingUnit;
+    }
+
+    public WaitingAnswer(T result) {
+        this();
+        this.result = result;
+    }
+
+    public WaitingAnswer(T result,
+                         long maxWaitingTime,
+                         TimeUnit maxWaitingUnit) {
+        this();
+        this.result = result;
+        this.maxWaitingTime = maxWaitingTime;
+        this.maxWaitingUnit = maxWaitingUnit;
+    }
+
+    /**
+     * Waits until answer is called in method {@link #answer(InvocationOnMock)}.
+     *
+     * @param maxWaitingTime
+     *         max time to wait
+     * @param maxWaitingUnit
+     *         time unit of the max waiting time argument
+     * @throws Exception
+     *         if the waiting time elapsed before this answer is called
+     * @see #answer(InvocationOnMock)
+     */
+    public void waitAnswerCall(long maxWaitingTime, TimeUnit maxWaitingUnit) throws Exception {
```

Comment:
I prefer just Exception. It shows that this method can throw exception and dev should understand that. It also allows to avoid catching and wrapping InterruptedException